### PR TITLE
Fix error serialization

### LINKED ISSUE #161: `Emitted value instead of an instance of Error` on WARNING under webpack 2.4.1
happypack 3.0.3
webpack 2.4.1

Likely this happens:
https://github.com/webpack/webpack/blob/e2bb2f41fa9f74f063043249dfc9c06ea6d08c03/lib/NormalModule.js#L113-L115
```js
class NonErrorEmittedError extends WebpackError {
	constructor(error) {
		super();

		this.name = "NonErrorEmittedError";
		this.message = "(Emitted value instead of an instance of Error) " + error;

		Error.captureStackTrace(this, this.constructor);
	}
}

// ...

	createLoaderContext(resolver, options, compilation, fs) {
		const loaderContext = {
			version: 2,
			emitWarning: (warning) => {
				if(!(warning instanceof Error))
					warning = new NonErrorEmittedError(warning);
				this.warnings.push(new ModuleWarning(this, warning));
			},
			emitError: (error) => {
				if(!(error instanceof Error))
					error = new NonErrorEmittedError(error);
				this.errors.push(new ModuleError(this, error));
			},
```

I've printed a stack of where this happens, and it's `happypack`:
```
       at Object.loaderContext.emitWarning (/__CENSORED__/node_modules/webpack/lib/NormalModule.js:112:71)
       at Object.LOADER_RPCS.emitWarning (/__CENSORED__/node_modules/happypack/lib/HappyRPCHandler.js:79:12)
       at HappyRPCHandler.execute (/__CENSORED__/node_modules/happypack/lib/HappyRPCHandler.js:56:22)
       at ChildProcess.acceptMessageFromWorker (/__CENSORED__/node_modules/happypack/lib/HappyThread.js:83:27)
       at emitTwo (events.js:87:13)
       at ChildProcess.emit (events.js:172:7)
       at handleMessage (internal/child_process.js:686:10)
       at Pipe.channel.onread (internal/child_process.js:440:11)
```

Found via `eslint-loader`: https://github.com/MoOx/eslint-loader/issues/172

Related: https://github.com/amireh/happypack/issues/152

### PR DESCRIPTION
**Pull request by waterfoul**
Fixes #161

### CODE REVIEW

**Comment by waterfoul** [comment]
@amireh I've got CI passing. What would be needed to get this merged?

**Comment by amireh** [comment]
Thanks for the contribution! I'm not sure about the implementation though, if the issue #161 is about emitWarning and emitError not emitting instances of Errors, this isn't the right way to solve it.

Also, we already have error (de)serializing routines in place that we use elsewhere to communicate errors between the background and the foreground, so for any solution to this problem it should be using those modules.

File: lib/HappyFakeLoaderContext.js [hunk 1]
```diff
@@ -1,6 +1,27 @@
 var resolveSyncNotSupported = NotSupported('this.resolveSync()');
 var execNotSupported = NotSupported('this.exec()');
 
+if (!('toJSON' in Error.prototype)) {
+  // Deliberately extending the Error prototype so that when node JSON.stringifys Errors
+  // they don't loose all their info
+  // eslint-disable-next-line no-extend-native
+  Object.defineProperty(Error.prototype, 'toJSON', {
```
Comment (amireh) [inline lib/HappyFakeLoaderContext.js]: extending native objects is generally regarded as bad practice, the right way to do this is to serialize the error to a JSON object prior to sending (that'd be in emitWarning and emitError on fake loader context) and then deserialize it back into an Error object on the receiving end

**Comment by amireh** [comment]
I've fixed this in https://github.com/amireh/happypack/commit/3dfc59f9e523b91019ce923ec3d83c53651ea85c so we can close this

thanks again! :)

**Pull request closed without merging**